### ISSUE TITLE
Change the behaviour of clean_names() when the input is a character vector

### ISSUE THREAD

**Issue by fernandovmacedo** [sfirke/janitor#197]
The function `clean_names()` is pretty useful, but it behaves (in my opinion) in an unexpected way when the input is a character vector. It generates a data.frame with all columns names empty and the character vector as the values of the first line.

The most conservative solution would be to throw an error if the input is not a data.frame, but I believe it would be more useful to return a character vector with "clean names" if the input is a character vector. In this way the function would useful in even more use case scenarios.

`clean_names` could become a generic function that calls methods according to the input type. For now would be character and data.frame, in the future other types could be implemented if needed. To avoid code repetition, internally the data.frame method could call the character method with names(input) as the argument. This could also be implemented without using S3 methods, using hidden functions, if turns out to create more difficulties then benefits.

To be even better (but maybe out of the scope of the package) the clean_names method for character vectors would have an additional argument `dupe_count = TRUE` that would enable the user to set it to false and avoid adding numbers to the column names. This would be useful because cleaning character vector would have others applications beside fixing data.frame columns names.

---
``` r
library(janitor)

clean_names(c("Type A","Type B","Type A","Type B", "Type C"))
#>     <NA>     <NA>     <NA>     <NA>     <NA> 
#> "Type A" "Type B" "Type A" "Type B" "Type C"
```

<details><summary>Session info</summary>

``` r
devtools::session_info()
#> Session info --------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.3.3 (2017-03-06)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  Portuguese_Brazil.1252      
#>  tz       America/Sao_Paulo           
#>  date     2018-05-01
#> Packages ------------------------------------------------------------------
#>  package    * version date       source                         
#>  assertthat   0.2.0   2017-04-11 CRAN (R 3.3.3)                 
#>  backports    1.0.5   2017-01-18 CRAN (R 3.3.2)                 
#>  bindr        0.1     2016-11-13 CRAN (R 3.3.3)                 
#>  bindrcpp     0.2     2017-06-17 CRAN (R 3.3.3)                 
#>  devtools     1.12.0  2016-06-24 CRAN (R 3.3.3)                 
#>  digest       0.6.9   2016-01-08 CRAN (R 3.3.1)                 
#>  dplyr        0.7.4   2017-09-28 CRAN (R 3.3.3)                 
#>  evaluate     0.10    2016-10-11 CRAN (R 3.3.3)                 
#>  glue         1.2.0   2017-10-29 CRAN (R 3.3.3)                 
#>  htmltools    0.3.5   2016-03-21 CRAN (R 3.3.1)                 
#>  janitor    * 1.0.0   2018-05-01 Github (sfirke/janitor@91f53f1)
#>  knitr        1.20    2018-02-20 CRAN (R 3.3.3)                 
#>  magrittr     1.5     2014-11-22 CRAN (R 3.3.1)                 
#>  memoise      1.0.0   2016-01-29 CRAN (R 3.3.1)                 
#>  pillar       1.2.2   2018-04-26 CRAN (R 3.3.3)                 
#>  pkgconfig    2.0.1   2017-03-21 CRAN (R 3.3.3)                 
#>  R6           2.2.2   2017-06-17 CRAN (R 3.3.3)                 
#>  Rcpp         0.12.16 2018-03-13 CRAN (R 3.3.3)                 
#>  rlang        0.2.0   2018-02-20 CRAN (R 3.3.3)                 
#>  rmarkdown    1.9     2018-03-01 CRAN (R 3.3.3)                 
#>  rprojroot    1.2     2017-01-16 CRAN (R 3.3.3)                 
#>  snakecase    0.9.1   2018-03-25 CRAN (R 3.3.3)                 
#>  stringi      1.1.1   2016-05-27 CRAN (R 3.3.0)                 
#>  stringr      1.2.0   2017-02-18 CRAN (R 3.3.3)                 
#>  tibble       1.4.2   2018-01-22 CRAN (R 3.3.3)                 
#>  withr        2.1.2   2018-03-15 CRAN (R 3.3.3)                 
#>  yaml         2.1.13  2014-06-12 CRAN (R 3.3.1)
```
</details>

**Comment by sfirke** [comment]
This is a thoughtful feature request, thank you.  I see how users would want to run `clean_names` on a vector and I think the proposed S3 implementation would be elegant.

The workhorse of `clean_names` is `snakecase::to_any_case()`, which works on character vectors.  It can leave duplicated strings as is, or append numbers.  I am a little wary of duplicating this function; whatever is in janitor won't touch the capability of this function, made for this purpose.

What do you think about printing an error message when `clean_names` is called on a vector saying, `clean_names runs on a data.frame.  To clean up a character vector, see snakecase::to_any_case` ?  Is there  a use case that `to_any_case` wouldn't cover?

**Comment by fernandovmacedo** [comment]
Your comments made me think of a little different solution. The problem is that `clean_names` do some additional transformations on the names, so results from `to_any_case` and `clean_names` are not equal.

My new suggestion would be to drop these additional transformations from `clean_names`  and put everything in `to_any_case`. To gives more flexibility to the user `clean_names` could gain a `...` argument to pass additional arguments to `to_any_case` directly from `clean_names`.

Take the question about numbering columns for example, `to_any_case` provide an option for the user to choose if it would want it or not (the default is no), but in `clean_names` the routine is coded in the function directly and after the call to `to_any_case`. In the end the user have no choice. But it would be possible to change this behaviour via a `...` argument passed to `to_any_case`. This implementation specifically could be trick because the default in snakecase is to not add the numbers. `clean_names` could have an argument `unique_sep = ""` passed to `to_any_case` or could test if the user have defined this argument, if not pass `unique_sep = ""` to `to_any_case` without putting the argument in the function formal arguments. (I don't know if I expressed this idea clearly).

Summing up, drop the transformations in `clean_names`, add a `...` argument to be passed to `to_any_case` and think of a solution to add numbers in the columns, it could be an argument `unique_sep = ""`. With this, the error message pointing to `to_any_case` would be good, as the results would be the same.

**Comment by jennybc** [comment]
I was about to open a new issue but will just weigh in here.

I think it would be good to offer the logic of `clean_names()` in a form that is "names in, names out" and then call that function from` clean_names()`, which is "named object in, re-named object out".

Context: I'm working on consistency around name repair in the tidyverse and possibly expanding the name repair controls in tibble. A key "escape valve" is likely to be: we will accept the user's name repair function in, e.g., `tibble()` or `as_tibble()`. Therefore, the function I'm requesting would be a valid candidate for that. Someone who loves `janitor::clean_names()` could apply that logic when they call `tibble()` or, eventually, `read_excel()` or `read_csv()`.

Full disclosure: we might also offer some in-house name repair that overlaps with some of what `clean_names()` does, but our version would likely do much less.

**Comment by Tazinho** [comment]
@sfirke If you aggree, I can submit a pullrequest changing `clean_names()` into a generic as @fernandovmacedo  suggested.

@jennybc regarding the disclosure: have you looked into the [snakecase pkg](https://github.com/Tazinho/snakecase) and its `to_any_case()` function?

It is underlying `clean_names()` and sounds like what you describe:
* its character in, character out
* it does much less than `clean_names()`. Basically `clean_names()` adds some preprocessing steps.
* but can do much more, since it is parameterized, modularized and designed for consistency.
* it has only stringr dependency (might be a downside)

**Comment by jennybc** [comment]
I'm a little leery of transforming `clean_names()` into a generic. It would seem better to implement in a regular function. Selfishly, I think that would work more robustly with the usage I want to make possible.

@Tazinho I just read up on snakecase! I can tell we are kindred spirits from the README 😀

We can't use it in tibble because we would never make tibble depend on stringi. So we'll have to offer some more basic options and then advertise more fully featured functions available elsewhere.

**Comment by jennybc** [comment]
Let me refine this:
 
>I'm a little leery of transforming clean_names() into a generic

Maybe that does make sense, but I'd generalize over "objects that can have names" as opposed to "objects with names" and "the names themselves".

**Comment by fernandovmacedo** [comment]
@jennybc I believe the function best suited for your case use is `to_any_case`, the only problem right now is that `to_any_case` does not have the same results as `clean_names`. With little changes `clean_names` could be almost a direct call to `to_any_case`. In this scenario `to_any_case` could be used with tibble and clean_names, generic or not, would be used to "clean the names of this thing". The arguments and outputs would be consistent across both scenarios.

**Comment by sfirke** [comment]
I am vacationing with blessedly poor internet connection so can't
participate in this in earnest for a bit, but wanted to encourage everyone
sharing ideas and discussing. This is a good idea and I think the design
and usage considerations will be much harder than the implementation. So
I'm thrilled to see you all hashing thile hard part out - thank you!

**Comment by billdenney** [comment]
From @jennybc:
> >I'm a little leery of transforming clean_names() into a generic
>
> Maybe that does make sense, but I'd generalize over "objects that can have names" as opposed to "objects with names" and "the names themselves".

I tend to agree with this.  I anticipate that users (myself included) would accidentally use this on a string vector and not get the expected result.  I could also imaging wanting to run something like the following to make string vector names prettier for later use (not that the below is especially pretty regardless):

``` r
> janitor::clean_names(c("(UgLyNaMe)"=1))
ug_ly_na_me 
          1 
> 
```

That functionality is lost if we make it operate sometimes on names of the object and sometimes on the value.

I'd prefer to see something like `clean_names_text()` or `make_usable_name()` (I'm not the best at naming functions :) ).  That would clarify that it is operating on the value of the object rather than the names of the object.  That function would take most of the work out of the middle of the current `clean_names()` but it would work on a logically different component of the input object.

**Comment by jennybc** [comment]
We are having a similar naming discussion in tibble/tidyverse and may try to delay the decision by not exporting any new functions for a while. It doesn't help that many of the relevant words ("clean", "tidy") are both adjective and verb.

Given the long pre-existence of `janitor::clean_names()` here, though, maybe `make_clean_names()` is the function that is "names in, names out"? At least it has a nod to `clean_names()` and `base::make.names()`.

**Comment by Tazinho** [comment]
@jennybc I agree on all points.

It might be best to leave `clean_names()` as it is, since it is well established, and:
* a generic with different behaviour changing names or elements is indeed a bad or at least a questionable idea,
* a generic changing only names of objects seems new and might be unexpected. It could confuse beginners, when they apply this function on a string.

I like `make_clean_names()`. Probably a better name than `to_clean_case()` or `to_janitor_case()`, which might have made sense from a "snakecase pkg perspective".

Regarding the stringr dependency: It might be possible to almost completely resolve this. Despite from possibly increasing implementation complexity two functionalities might be hard to handle via base R:
1. out of the box vectorization of most arguments
2. transliterations:
  * these would still be possible (also in combination):
```
to_any_case("Doppelgänger", transliterations = c("ä" = "ae", "ü" = "oe"))
#> [1] "doppelgaenger"
to_any_case("Doppelgänger", transliterations = "german")
#> [1] "doppelgaenger"
```

  * this one not (since it supplies a transliteration option from `stringi::stri_trans_list()` to `stringi::stri_trans_general()`):
```
to_any_case("Doppelgänger", transliterations = "Latin-ASCII")
#> [1] "doppelganger"
```

However, not sure how much you want to go into this direction.

**Comment by jennybc** [comment]
We are likely to offer some very basic name repair that's easy to do with lean dependencies and leave it to the user, who has full access to janitor and snakecase, to opt-in to more involved transformations.

**Comment by Tazinho** [comment]
I just looked into the tibble name-repair branch. Looks quite nice for snakecase, see below. Would definitely make sense for janitor also.

``` r
# devtools::install_github("tidyverse/tibble", ref = "name-repair")
library(purrr)
library(tibble)
library(snakecase)

as_tibble(iris, .name_repair = to_snake_case)
#> # A tibble: 150 x 5
#>    sepal_length sepal_width petal_length petal_width species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # ... with 140 more rows

iris %>% set_names(to_snake_case) %>% head()
#>   sepal_length sepal_width petal_length petal_width species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
#> 6          5.4         3.9          1.7         0.4  setosa
```

Created on 2018-09-14 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by jennybc** [comment]
@Tazinho That has been merged into master now, so you should be able to do the above by installing dev tibble from master. Thanks for experimenting!

**Comment by sfirke** [comment]
Putting this all into action: seems like the consensus is to:

1. Create a function `make_clean_names` that takes a character vector and returns a character vector.  This will use the code currently in `clean_names`.
2. Rewrite `clean_names` to call `make_clean_names` to avoid redundancy.
3. Update tests accordingly.

I like `make_clean_names` for the reasons @jennybc provides but if anyone has another name idea, please suggest.

Would anyone like to create a PR implementing this?  It would be nice to get this up on CRAN before the current dev version of tibble goes to CRAN.  @jennybc, do you have a goal date in mind for that?

**Comment by jennybc** [comment]
I think the tibble release is planned to mid- to late October.

**Comment by sfirke** [comment]
Thanks @Tazinho for implementing `make_clean_names`!  This now works on the dev version of tibble:
`as_tibble(iris, .name_repair = janitor::make_clean_names)`
